design/aws-ecs/aws-ecs-multiple-target-groups-support.md

```diff
@@ -0,0 +1,641 @@
+      }
+      const props = addProps as elbv2.AddApplicationTargetsProps;
+      targetGroup = new elbv2.ApplicationTargetGroup(this.stack, id + "Group", {
+        deregistrationDelay: props.deregistrationDelay,
+        healthCheck: props.healthCheck,
+        port: props.port,
+        protocol: props.protocol,
+        slowStart: props.slowStart,
+        stickinessCookieDuration: props.stickinessCookieDuration,
+        targetGroupName: props.targetGroupName,
+        vpc: listener.loadBalancer.vpc
+      });
+
+      (listener as elbv2.ApplicationListener).addTargetGroups(id, {
+        targetGroups: [targetGroup],
+        hostHeader: props.hostHeader,
+        pathPattern: props.pathPattern,
+        priority: props.priority
+      });
+    } else if (listener instanceof elbv2.NetworkListener) {
+      if (!listener.loadBalancer.vpc) {
+        // tslint:disable-next-line:max-line-length
+        throw new Error('Can only call registerContainerTarget() when using a Load Balancer with VPC');
+      }
+      const props = addProps as elbv2.NetworkTargetGroupProps;
+      targetGroup = new elbv2.NetworkTargetGroup(this.stack, id + "Group", {
+        deregistrationDelay: props.deregistrationDelay,
+        healthCheck: props.healthCheck,
+        port: props.port,
+        proxyProtocolV2: props.proxyProtocolV2,
+        targetGroupName: props.targetGroupName,
+        vpc: listener.loadBalancer.vpc
+      });
+      (listener as elbv2.NetworkListener).addTargetGroups(id, targetGroup);
+    } else {
+      throw new Error('Listener can only be either ApplicationListener or NetworkListener.');
+    }
+    return targetGroup;
+  }
+
+  ...
+
+  private attachToELB(targetGroup: elbv2.ITargetGroup, target: ContainerTarget): elbv2.LoadBalancerTargetProps {
+    if (this.taskDefinition.networkMode === NetworkMode.NONE) {
+      throw new Error("Cannot use a load balancer if NetworkMode is None. Use Bridge, Host or AwsVpc instead.");
+    }
+
+    this.loadBalancers.push({
+      targetGroupArn: targetGroup.targetGroupArn,
+      containerName: target.containerName,
+      containerPort: target.containerPort,
+    });
+
+    // Service creation can only happen after the load balancer has
+    // been associated with our target group(s), so add ordering dependency.
+    this.resource.node.addDependency(targetGroup.loadBalancerAttached);
+
+    const targetType = this.taskDefinition.networkMode === NetworkMode.AWS_VPC ? elbv2.TargetType.IP : elbv2.TargetType.INSTANCE;
+    return { targetType };
+  }
+
+  ...
+
+}  
+```
+
+### Part 4: Improve the current API
+
+``` ts
+/**
+ * The base class for Ec2Service and FargateService services.
+ */
+export abstract class BaseService extends Resource
+  implements IService, elbv2.IApplicationLoadBalancerTarget, elbv2.INetworkLoadBalancerTarget {
+
+  ...
+
+  private targetContainers?: ContainerTarget;
+
+  ...
+
+  protected registerTargetBase(target: ContainerTarget) {
```

Comment:
Depending on the order of operations, use case #2 would either effectively make `this.targetContainers` a scratchpad that updates on each target registration (if depth first) or not work properly (if both registrations happen before both attachments).  Do you know which is the case?

If the ordering is register/attach/register/attach I would suggest an annotation that the property is unstable/transient.

Reply:
I think the first one will be the case. However, I do agree that if we are going to implement this method then it should be annotated.

BTW, do you have any question or concern about the new API?

Reply:
Added thoughts below :)